pkg/internalregistry/legacy.go

```diff
@@ -0,0 +1,173 @@
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	applycorev1 "k8s.io/client-go/applyconfigurations/core/v1"
+	"k8s.io/client-go/util/csaupgrade"
+	"sigs.k8s.io/structured-merge-diff/v4/fieldpath"
+)
+
+func (c *imagePullSecretsController) legacyImagePullSecretName(ctx context.Context, serviceAccount *corev1.ServiceAccount) string {
+	// find the legacy image pull secret in the same namespace
+	secrets, err := c.secretLister.Secrets(serviceAccount.Namespace).List(labels.Everything())
+	if err != nil {
+		runtime.HandleError(err)
+		return ""
+	}
+	for _, secret := range secrets {
+		if isLegacyImagePullSecretForServiceAccount(secret, serviceAccount) {
+			return secret.Name
+		}
+	}
+	return ""
+}
+
+var legacyAnnotations = []string{
+	corev1.ServiceAccountNameKey,
+	corev1.ServiceAccountUIDKey,
+	controllers.ServiceAccountTokenSecretNameKey,
+	controllers.ServiceAccountTokenValueAnnotation,
+}
+
+func removeLegacyAnnotations(secret *corev1.Secret) {
+	for _, a := range legacyAnnotations {
+		delete(secret.Annotations, a)
+	}
+}
+
+var expectedLegacyAnnotations = map[string]func(*corev1.ServiceAccount, string) bool{
+	corev1.ServiceAccountNameKey:                   func(sa *corev1.ServiceAccount, v string) bool { return sa.Name == v },
+	corev1.ServiceAccountUIDKey:                    func(sa *corev1.ServiceAccount, v string) bool { return sa.UID == types.UID(v) },
+	controllers.ServiceAccountTokenSecretNameKey:   func(sa *corev1.ServiceAccount, v string) bool { return true },
+	controllers.ServiceAccountTokenValueAnnotation: func(sa *corev1.ServiceAccount, v string) bool { return true },
+}
+
+func isLegacyImagePullSecretForServiceAccount(secret *corev1.Secret, serviceAccount *corev1.ServiceAccount) bool {
+	for key, valueOK := range expectedLegacyAnnotations {
+		value, ok := secret.Annotations[key]
+		if !ok {
+			return false
+		}
+		if !valueOK(serviceAccount, value) {
+			return false
+		}
+	}
+	return true
+}
+
+func (c *imagePullSecretsController) cleanupLegacyResources(ctx context.Context, serviceAccount *corev1.ServiceAccount) error {
+
+	// get the image pull secret
+	imagePullSecret, err := c.secretLister.Secrets(serviceAccount.Namespace).Get(serviceAccount.Annotations[InternalRegistryImagePullSecretRefKey])
+	if err != nil {
+		return err
+	}
+
+	//  delete the legacy token secret
+	err = c.cleanupLegacyTokenSecret(ctx, imagePullSecret)
+	if err != nil {
+		return err
+	}
+
+	// take ownership of image pull secret legacy annotations
+	imagePullSecret, err = c.cleanupLegacyFieldOwnership(ctx, imagePullSecret)
+	if err != nil {
+		return err
+	}
+
+	// remove image pull secret legacy annotations
+	imagePullSecret, err = c.cleanupLegacyImagePullSecretAnnotations(ctx, imagePullSecret)
+	if err != nil {
+		return err
+	}
+
+	// remove the image pull secret from the service account's mountable secrets
+	c.cleanupLegacySecretRefs(ctx, imagePullSecret.Name, serviceAccount)
```

Comment:
I misunderstood what the [secrets refs were for](https://kubernetes.io/docs/reference/kubernetes-api/authentication-resources/service-account-v1/#:~:text=secrets%20([]ObjectReference)). I do not think I should be removing it.